### PR TITLE
drivers: ns16550: Implement RX related features

### PR DESCRIPTION
**Pull request by gagachang**
Implement ns16550_getchar() and ns16550_have_rx_data() for RX related serial operations into ns16550 UART driver.

### CODE REVIEW

File: core/drivers/ns16550.c [hunk 1]
```diff
@@ -70,9 +71,38 @@ static void ns16550_putc(struct serial_chip *chip, int ch)
 	serial_out(base + (UART_THR << pd->reg_shift), pd->io_width, ch);
 }
 
+static bool ns16550_have_rx_data(struct serial_chip *chip)
+{
+	struct ns16550_data *pd =
+		container_of(chip, struct ns16550_data, chip);
+	vaddr_t base = io_pa_or_va(&pd->base,
+				   (UART_LSR << pd->reg_shift) + pd->io_width);
```
Comment (jenswi-linaro) [inline core/drivers/ns16550.c:79]: There's quite a bit of repeated code here, how about adding a `chip_to_base()` function like in the pl011.c driver and refactor a bit?

Comment (gagachang) [reply core/drivers/ns16550.c:79]: Good suggestion, I'll try it

Comment (gagachang) [reply core/drivers/ns16550.c]: It seems there are some configurable parameters included in `struct ns16550_data *pd`
We might not able to eliminate it in every function

**Comment by gagachang** [comment]
Implement `chip_to_base()` for ns16550.

**Comment by gagachang** [comment]
Implement `chip_to_base_and_data()` to get vaddr and private structure in one function call.

File: core/drivers/ns16550.c [hunk 2]
```diff
@@ -45,12 +45,18 @@
...
+	return io_pa_or_va(&(*pd)->base, NS16550_UART_REG_SIZE);
+}
+
 static void ns16550_flush(struct serial_chip *chip)
 {
-	struct ns16550_data *pd =
-		container_of(chip, struct ns16550_data, chip);
-	vaddr_t base = io_pa_or_va(&pd->base,
-				   (UART_LSR << pd->reg_shift) + pd->io_width);
+	struct ns16550_data *pd;
```
Comment (etienne-lms) [inline core/drivers/ns16550.c:59]: = NULL

File: core/drivers/ns16550.c [hunk 3]
```diff
@@ -59,10 +65,8 @@ static void ns16550_flush(struct serial_chip *chip)
 
 static void ns16550_putc(struct serial_chip *chip, int ch)
 {
-	struct ns16550_data *pd =
-		container_of(chip, struct ns16550_data, chip);
-	vaddr_t base = io_pa_or_va(&pd->base, (UART_THR << pd->reg_shift) +
-					       pd->io_width);
+	struct ns16550_data *pd;
```
Comment (etienne-lms) [inline core/drivers/ns16550.c:69]: ditto

File: core/drivers/ns16550.c [hunk 4]
```diff
@@ -74,9 +75,34 @@ static void ns16550_putc(struct serial_chip *chip, int ch)
 	serial_out(base + (UART_THR << pd->reg_shift), pd->io_width, ch);
 }
 
+static bool ns16550_have_rx_data(struct serial_chip *chip)
+{
+	struct ns16550_data *pd;
```
Comment (etienne-lms) [inline core/drivers/ns16550.c:80]: = NULL

File: core/drivers/ns16550.c [hunk 5]
```diff
@@ -74,9 +75,34 @@ static void ns16550_putc(struct serial_chip *chip, int ch)
...
+	vaddr_t base = chip_to_base_and_data(chip, &pd);
+
+	return serial_in(base + (UART_LSR << pd->reg_shift), pd->io_width) &
+	       UART_LSR_DR;
+}
+
+static int ns16550_getchar(struct serial_chip *chip)
+{
+	struct ns16550_data *pd;
```
Comment (etienne-lms) [inline core/drivers/ns16550.c:89]: ditto

File: core/drivers/ns16550.c (hunk 2 above)
Comment (gagachang) [reply core/drivers/ns16550.c]: resolved

**Review by jenswi-linaro — COMMENTED** [review]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Pull request merged**